### ISSUE TITLE
Transactional writer got stuck after receiving SEALED_SEGMENT reply 

### ISSUE THREAD

**Issue by RaulGracia** [pravega/pravega#4411]
**Problem description**
I have triggered a longevity experiment (build 194) that has failed with Pravega 0.6 RC5. The workload is `small-scale-tx`, which consist of exercising a transactional writer on a Stream with scaling enabled, plus a reader.

The problem with this workload happened after 1 hour, as follows:
1. The controller performs a scale operation: 
```
2019-11-29 10:34:50,306 5426385 [eventprocessor-10] INFO  i.p.c.s.e.r.AutoScaleTask - [requestId=114044926125503922] Scale up request received for stream segment longevity/smallScaleTx/15.#epoch.20
2019-11-29 10:34:50,330 5426409 [eventprocessor-3] INFO  i.p.c.s.e.r.StreamRequestHandler - Processing scale request 114044926125503922 for stream longevity/smallScaleTx
2019-11-29 10:34:50,333 5426412 [eventprocessor-18] INFO  i.p.c.s.e.r.ScaleOperationTask - [requestId=114044926125503922] starting scale request for longevity/smallScaleTx segments [85899345935] to new ranges [0.25=0.28125, 0.28125=0.3125]
2019-11-29 10:34:50,358 5426437 [epollEventLoopGroup-2-3] INFO  i.p.controller.server.SegmentHelper - [requestId=114044926125503922] CreateSegment longevity/smallScaleTx/35.#epoch.26 SegmentCreated 2995692444319744.
2019-11-29 10:34:50,366 5426445 [epollEventLoopGroup-2-3] INFO  i.p.controller.server.SegmentHelper - [requestId=114044926125503922] CreateSegment longevity/smallScaleTx/36.#epoch.26 SegmentCreated 2995696739287040.
2019-11-29 10:34:50,412 5426491 [epollEventLoopGroup-2-3] INFO  i.p.controller.server.SegmentHelper - [requestId=114044926125503922] SealSegment longevity/smallScaleTx/15.#epoch.20 SegmentSealed 2995756868829184.
2019-11-29 10:34:50,477 5426556 [eventprocessor-11] INFO  i.p.c.s.e.r.ScaleOperationTask - [requestId=114044926125503922] scale processing for longevity/smallScaleTx epoch 25 completed.
2019-11-29 10:34:50,477 5426556 [eventprocessor-11] INFO  i.p.c.s.e.r.ScaleOperationTask - [requestId=114044926125503922] scale request for longevity/smallScaleTx segments [85899345935] to new ranges [0.25=0.28125, 0.28125=0.3125] completed successfully.
2019-11-29 10:34:50,479 5426558 [eventprocessor-1] INFO  i.p.c.s.e.r.StreamRequestHandler - Processing scale request 114044926125503922 for stream longevity/smallScaleTx complete
```
2. One of the created Segments is this `35.#epoch.26`:
```
2019-11-29 10:34:50,354 5359614 [epollEventLoopGroup-11-3] INFO  i.p.s.s.h.h.PravegaRequestProcessor - [requestId=2995692444319744] Creating stream segment WireCommands.CreateSegment(type=CREATE_SEGMENT, requestId=2995692444319744, segment=longevity/smallScaleTx/35.#epoch.26, scaleType=2, targetRate=3).
```
3. But this segment gets sealed immediately after being created:
```
2019-11-29 10:34:50,761 5360021 [epollEventLoopGroup-11-8] INFO  i.p.s.s.h.h.PravegaRequestProcessor - [requestId=2997191387906048] Sealing segment WireCommands.SealSegment(type=SEAL_SEGMENT, requestId=2997191387906048, segment=longevity/smallScaleTx/35.#epoch.26) 
```
4. At the same time, the longevity application writer fetches this segment to start writing to it:  
```
2019-11-29 10:34:50,780 4213987 [clientInternal-1-7] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Fetching endpoint for segment longevity/smallScaleTx/35.#epoch.26, writer 926f3921-beb9-44d3-bf58-e067f8fc99bd
```
5. It establishes the connection against the Segment Store:
```
2019-11-29 10:34:50,781 4213988 [clientInternal-1-1] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Establishing connection to PravegaNodeUri(endpoint=172.28.75.13, port=12345) for longevity/smallScaleTx/35.#epoch.26, writerID: 926f3921-beb9-44d3-bf58-e067f8fc99bd
```
6. But immediately after it receives a SEGMENT_SEALED message from the Segment Store:
```
2019-11-29 10:34:50,802 4214009 [epollEventLoopGroup-2-16] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Received SegmentSealed WireCommands.SegmentIsSealed(type=SEGMENT_IS_SEALED, requestId=352118598795264, segment=longevity/smallScaleTx/35.#epoch.26, serverStackTrace=, offset=-1) on writer 926f3921-beb9-44d3-bf58-e067f8fc99bd
```
7. After that, the longevity writer does not make any progress and seems completely stuck:
```
2019-11-29 10:35:13,293 4236500 [pool-2-thread-1] INFO  i.p.l.utils.PerformanceUtils - TASK  Running for 1h 10m
2019-11-29 10:35:13,293 4236500 [pool-2-thread-1] INFO  i.p.l.utils.PerformanceUtils - Writers (1/1): events:178,949, events/sec:42, KB/sec:32.435547
2019-11-29 10:35:13,293 4236500 [pool-2-thread-1] INFO  i.p.l.utils.PerformanceUtils -  Transactions: started: 11960, perSec:1.4052009456264776, committed:5944, failed:0, aborted:6016, aborted but read: 0
...
2019-11-29 21:24:43,293 43206500 [pool-2-thread-1] INFO  i.p.l.utils.PerformanceUtils - TASK  Running for 12h 0m
2019-11-29 21:24:43,293 43206500 [pool-2-thread-1] INFO  i.p.l.utils.PerformanceUtils - Writers (1/1): events:178,949, events/sec:4, KB/sec:3.1757812
2019-11-29 21:24:43,293 43206500 [pool-2-thread-1] INFO  i.p.l.utils.PerformanceUtils -  Transactions: started: 11960, perSec:0.1375925925925926, committed:5944, failed:0, aborted:6016, aborted but read: 0
```
That is, there are no more writer-related messages in the longevity app, looks like the writer is alive from the application's perspective (i.e., `Writers (1/1)`), but completely blocked (no more transactions created, not even stack traces).

**Problem location**
Interactions client/segment store?

**Suggestions for an improvement**
Understand why the writer stopped making progress and fix the issue.

**Comment by shiveshr** [comment]
So segments being created and sealed immediately is a result of rolling transaction where multiple segments are sealed and their duplicates are created. 

What's curious is that *Transaction writers* do not deal with regular segments. They deal with transaction segments which are completely decoupled from regular segments. 
A transaction writer should only deal with segments where the segment name follows transaction's naming convention. 
But here, for the application's writers to even be aware of regular segments of the stream is unusual. 

The writer has received segment sealed for a bunch of segments. 
```
2019-11-29 10:34:50,802 4214009 [epollEventLoopGroup-2-16] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Received SegmentSealed WireCommands.SegmentIsSealed(type=SEGMENT_IS_SEALED, requestId=352118598795264, segment=longevity/smallScaleTx/35.#epoch.26, serverStackTrace=, offset=-1) on writer 926f3921-beb9-44d3-bf58-e067f8fc99bd
```
Upon receiving these segment sealed exceptions, it has not even requested controller for their successors which is what a regular writer would have done. 

So in all i see two issues:
1. Transaction writers being aware of regular segments and setting up appends on them. 
2. Upon receiving segment sealed, they do not make any progress (which may be the write behavior for a transaction writer anyway). But it should have thrown some exception and exited letting the application know something bad had transpired.

**Comment by tkaitchuck** [comment]
I assume this is a result of using the Depricated `EventStreamWriter.beginTxn()` api and using this transaction as a transactional writer. As this API has been deprecated for a full year now, I think we can simply delete the implementation and revolve it.

**Comment by tkaitchuck** [comment]
I have created https://github.com/pravega/pravega/issues/4416 to track this.

**Comment by RaulGracia** [comment]
Fixed via #4416.
